docs/proposals/23-autoscaling-kube-apiserver-via-independent-hpa-and-vpa.md

```diff
@@ -0,0 +1,347 @@
+ShootKapi, and then would reduce the response to only `apiserver_request_total` counters, before passing it on to the
+caller). 
+
+### Element: New Custom Pod Metric for ShootKapis
+A new `shoot:apiserver_request_total:rate` pod custom metric is made available for each ShootKapi pod on the
+seed. It is provided by the gardener-custom-metrics component. It is the rate of API requests per second, broken down
+by ShootKapi pod.
+
+### Element: HPA
+ShootKapi is scaled horizontally via HPA operating on the aforementioned average kube-apiserver request rate custom
+metric. Scale-in is delayed by a stabilization window. This avoids unnecessary flapping, the effects of which would
+be further amplified by VPA-driven evictions, and serves as a general precaution against HPA-VPA resonance,
+by ensuring that the two autoscalers operate with different frequency.
+
+The scaling threshold value for HPA is initially set to a conservative (lower-biased) estimate of the requests/replica
+ratio observed with the existing solution. It will be then gradually fine-tuned to the maximum value which does not
+impact quality of service.
+
+This proposal is the first stage of an incremental, two stage approach. It aims primarily at resolving the existing
+stability issues inherent to HVPA and enabling (but not executing) scaling efficiency optimizations. The second stage
+will be focused solely on further fine-tuning scaling efficiency to optimize ShootKapi compute cost over HVPA.
+Details follow in the [Discussion and Limitations](#discussion-and-limitations) section.
+
+### Element: VPA
+A standard VPA operates on ShootKapi. Its primary purpose is to improve resource efficiency over what's possible via
+HPA's coarse, replica-granularity scaling, by shrinking replicas down to match actual utilization.
+
+This improves over the existing solution by having vertical scaling active the entire time and not only during certain
+periods.
+
+### Transition Strategy
+The following steps will be executed over time, and provide a non-disruptive transition from the existing implementation
+to the one hereby proposed.
+
+1. Proposed scaling approach will be deployed behind a feature gate and disabled by default. Existing HVPA feature flag
+   will be preserved at this point, as its use is not limited to kube-apiserver scaling.
+2. New feature gate gradually enabled until all ShootKapi workloads scaled by new approach (Note: this step applies to
+   individual garden/seed instances, and not to the code base as a whole).
+3. At this point, the preexisting HVPA feature flag has no effects on ShootKapi scaling (Note: this step applies to
+   individual garden/seed instances, and not to the code base as a whole).
+4. New feature flag promoted and eventually removed.
+5. (Outside the scope of this proposal) Use of HVPA removed for other workloads and the preexisting HVPA flag removed.
```

Comment:
I like the step-by-step approach. How do we validate the effects of step 2 and how do we take the decision to roll-back or go on with the further deployment/enablement?

Autoscaling has always 2 opposing goals:
- Have enough capacity to ensure availability, latency, etc.
- Have not too much capacity provisioned, so that we don't waste money on superfluous resources

The decision should incorporate both sides, right?

Reply:
>How do we validate the effects of step 2

That question has been standing before us from the very start, and still doesn't have an ideal answer. To be representative, measurements need to be against a real-life system. But on a real life system, there are major uncontrolled variables - most importantly, other software changes get applied in the same time frame.

Some obvious candidate measures are the total requested resources, and actual resource use. In practice, it's also tempting to compare seed nodes before and after the change.

In a way, we already have that measurement, as we recently experimented with removing class-based scaling - a removal which is a big part of the efficiency gain of the proposed approach. IIRC, depending on the specific configuration of the clusters, that change can result in more than 30% reduction in the number of seed nodes. So a big part of the efficiency gain of the proposed approach is already available in the code base.

As to stability, the metrics are the ones you mention. Some are consumer level metrics, such as latency, requests rejections. There may also be some internal metrics worth considering, like eviction count. Overall, I think the evaluation will be exploratory (as opposed to applying a previously designed metric). This approach is vulnerable to observer bias, but defining one overall metric for autoscaling quality is no trivial matter: if one autoscaler is better at X, and another at Y, how to we weigh X and Y against each other. Is a 10 second latency event twice as bad, as a 5 second one, or should we model e.g. a squared cost. In the end, I don't expect that we'll come up with one objective metric. Instead, we'll document what measurements we made, and what results we got, and hope that we all agree on the interpretation of the results.

Reply:
While your concerns may be justified (other changes that possibly influence availability or utilisation positively or negatively in parallel to the roll-out of your changes), I would put them aside. They are no reason not to look (in real life, everything happens in parallel). Also, the likelihood is very low, that parallel changes will effect the same areas (here availability and/or utilisation), even on the same seed.

So, that concern aside, what concrete "measurement instruments" will you employ? You iterated a few metrics, but are you planning on checking all of them and how? What you could leverage are the landscape probes for availability and usage (see https://github.tools.sap/kubernetes/gardener-robot/tree/master/robot/garden/probes/landscape). The detailed report also has the numbers per seed, I believe. However, this is a very crude and primitive instrument (a hack to have some minimal insights) - maybe you like to talk with @istvanballok for something smarter/better? Or discuss with the landscape stability team, e.g. @dguendisch and/or @BeckerMax?

Reply:
The GEP now specifies a list of quantities to be used when evaluating the fitness of the new autoscaling solution.

>already developed probes/instruments that we could use (...)?

First, thanks for the pointer! On your question: nothing worth mentioning. I mostly manually explored available data, with the occasional throwaway automation script. I have a garden crawler tool I use, but it has far fewer features than the probes, and I only favour it because it's built on technology with which I'm efficient.